### ISSUE TITLE
Error on explicit `any` parameter with noImplicitAny enabled

### ISSUE THREAD

**Issue by haizz** [Microsoft/TypeScript#12822]
**TypeScript Version:**  2.1.4
**Code**

```ts
// compile with --noImplicitAny switch
function f1(props: any): any {
    return {
        a: [], 
        ...props
    };
}

function f2(data: any): void {
};
```

**Expected behavior:**

Code is compiled without errors.

**Actual behavior:**

```
8 function f2(data: any): void {
              ~~~~~~~~~

test.ts(8,13): error TS7006: Parameter 'data' implicitly has an 'any' type.
```

These variants are compiled fine for some weird reason:

```ts
function f1(props: any): any {
    return {
        a: 1, 
        ...props
    };
}

function f2(data: any): void {
};
```

```ts
function f1(props: any): any {
    return {
        a: [], 
    };
}

function f2(data: any): void {
};
```

**Comment by RyanCavanaugh** [comment]
Shortest repro I could get. This is the weirdest bug I've seen in a while!
```ts
const props = <any>null;
console.log({ a: [], ...props });
declare function f2(data: any): void;
```

**Comment by haizz** [comment]
**Possible workaround:**

Cast `[]` to specific type:

```typescript
// this works as intended
function f1(props: any): any {
    return {
        a: [] as number[] /* or any[] */,
        ...props
    };
}

function f2(data: any): void {
};
```

The problem can be tricky to identify because single `a: []` line in one file breaks compilation of dozen files all over the project.

**Comment by sandersn** [comment]
Another wonky workaround: turn on strictNullChecks.

**Comment by sandersn** [comment]
`checkObjectLiteral` assumes that the return value of `getSpreadType` is a fresh object type, but this is not true when spreading an expression of type `any`. So it incorrectly adds flags to `any` that it inherits from the rest of the object literal. 

In this case, `[]` contributes a ContainsWideningType flag, which makes all uses of `any` trigger an implicit any error.

**Comment by sandersn** [comment]
Fix is up at #12896
